File: go/034.go
package main

/**
在排序数组中查找元素的第一个和最后一个位置
思路：二分找到元素，往后推出右边坐标即可
*/
func searchRange(nums []int, target int) []int {
	l, r, err := 0, len(nums)-1, []int{-1, -1}
	if r == -1 {
		return err
	}
	for l < r {
		mid := (l + r) >> 1
		if nums[mid] >= target {
			r = mid
		} else {
			l = mid + 1
		}
	}
	if nums[l] != target {
		return err
	}
	var i int
	for i = l; i < len(nums) && nums[i] == nums[l]; i++ {
	}
	return []int{l, i - 1}
}

File: c++/034.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l=0,h=nums.size()-1,m,r1=-2,r2=0;
        // if(h==0)
        //     return {-1,-1};
        while(l<=h){
            m = (l+h)/2;
            if(nums[m] == target){
                r1 = r2 = m;
                break;
            }else if(nums[m] < target){
                l = m+1;
            }else{
                h = m-1;
            }
        }
        if(l<=h){
            while(r1>=0&&nums[r1]==target)
                --r1;
            while(r2<nums.size()&&nums[r2]==target)
                ++r2;
        }
        return {r1+1,r2-1};
    }
};